### PR TITLE
Adding support for ARB (Application Resource Bundle) (.arb) format

### PR DESCRIPTION
**Pull request by dsavinov-actionengine**
Problem and/or solution
-----------------------
Adding parsing and compiling for ARB (Application Resource Bundle) (.arb) format

How to test
-----------
### 1. Running unit-tests
`/openformats/tests/formats/arb/test_arb.py` contains tests for arb
Use `pytest /openformats/tests/formats/arb/test_arb.py` to run tests

### 2. Through testbed
Use "ARB" handler in the testbed

Reviewer checklist
------------------

Code:
* [ ] Change is covered by unit-tests
* [ ] Code is well documented, well styled and is following [best practices](https://tem.transifex.com)
* [ ] Performance issues have been taken under consideration
* [ ] Errors and other edge-cases are handled properly

PR:
* [ ] Problem and/or solution are well-explained
* [ ] Commits have been squashed so that each one has a clear purpose
* [ ] Commits have a proper commit message [according to TEM](https://tem.transifex.com/github-guide.html#working-on-a-feature)

### CODE REVIEW

**Comment by dsavinov-actionengine** [comment]
Tagging @kbairak for review, but feel free to tag others if needed

File: openformats/formats/json.py [hunk 1]
```diff
@@ -472,6 +472,123 @@ def unescape(string):
...
+                    self.transcriber.copy_until(key_position)
+                    raise ParseError(u"Duplicate string key ('{}') in line {}".
+                                     format(key, self.transcriber.line_number))
+                if nest is None:  # store all root-level keys in order to detect duplication
+                    self.existing_keys.add(key)
+                elif key.startswith("@") and key.endswith(".type") and value != "text":
+                    self.existing_keys.add(key)
+
+                if isinstance(value, DumbJson):
+                    self._find_keys(value, key)
```
Comment (kbairak) [inline openformats/formats/json.py:523]: Instead of going infinitely deep, we could instead search up to level two and collect entries that are non-text. This way we don't have to bother with escaping keys etc. Here is an idea; I haven't tested it so there might be a few errors:

```python
class ArbHandler(JsonHandler):
    ...

    def parse(self, content, **kwargs):
        ...

        keys = self._find_keys(parsed)
        stringset = self._extract(parsed, keys)

    def _find_keys(self, parsed):
        if parsed.type != dict:
            raise ParseError("...")

        keys, keys_to_ignore = set(), set()
        for key, key_position, value, _ in parsed:
            if not key.startswith("@"):
                if not isinstance(value, str):
                    continue
                if key in keys:
                    raise ParseError(
                        f"Duplicate string key ('{key}') in line "
                        f"{self.transcriber.line_number}"
                    )
                keys.add(key)
            else:
                if isinstance(value, DumbJson):
                    (inner_value, _), = value.find_children('type')
                    if inner_value == "text":
                        keys_to_ignore.add(key)
        return keys - keys_to_ignore

    def _extract(self, parsed, keys):
        stringset = []
        for key, _, value, value_position in parsed:
            if key not in keys:
                continue
            ...
        return stringset
```

File: openformats/formats/json.py [hunk 2]
```diff
@@ -472,6 +472,123 @@ def unescape(string):
...
+                                                         value_position)
+                    if openstring:
+                        self.stringset.append(openstring)
+                else:
+                    # Ignore other JSON types (bools, nulls, numbers)
+                    pass
+        else:
+            raise ParseError("Invalid JSON")
+
+    def compile(self, template, stringset, **kwargs):
```
Comment (kbairak) [inline openformats/formats/json.py:553]: So, the only reason you re-define `compile` and `_copy_until_and_remove_section` is so that you can add the `keep_sections` functionality, which is only needed for testing. I am not very fond of this idea. It makes the code unnecessarily complex. However, there is a case where not removing the sections makes sense.

This is not adequately described in the docs, but when Transifex invokes this handler, it will pass `is_source` as a keyword argument. In this case, we can be certain that

1. the stringset will not be missing anything and that
2. the user wants to get the exact same file they previously uploaded

So, if this is indeed the same use-case, maybe you can rename `keep_sections` to `is_source`.

In any case, lets make sure that when `is_source=True` is passed to `compile`, the end result is the same as the originally uploaded file.

File: openformats/formats/json.py [hunk 3]
```diff
@@ -472,6 +472,123 @@ def unescape(string):
...
+
+                if isinstance(value, DumbJson):
+                    self._find_keys(value, key)
+                else:
+                    pass
+        else:
+            raise ParseError("Invalid JSON")
+
+    def _extract(self, parsed):
+        if parsed.type == dict:
```
Comment (kbairak) [inline openformats/formats/json.py:530]: No need for this check, we already did it in `_find_keys`. Perhaps to make it clearer, lets make one check in `parse` and then in `_find_keys` and `_extract` assume that `parsed` represents a dict.

**Review by kbairak — CHANGES_REQUESTED** [review]
Great job!

**Comment by kbairak** [comment]
There is another issue I am concerned about. Looking at the examples of the [official Flutter documentation](https://docs.flutter.dev/ui/accessibility-and-internationalization/internationalization#setting-up), it looks like the language files are supposed to be more bare: only key-translation pairs without any metadata.

image

So, the question is: should we follow this? should we make sure that when `is_source=False` during `compile`, only the key-translation pairs should be present while when `is_source=True` we should make sure that they get back the same file they originally uploaded?

Have you looked into this?

File: openformats/formats/json.py (hunk 1 above)
Comment (dsavinov-actionengine) [reply openformats/formats/json.py]: Did not try this. Anyway, ARBs normally have only two levels of json, so we don't expect infinite nesting.

File: openformats/formats/json.py (hunk 2 above)
Comment (dsavinov-actionengine) [reply openformats/formats/json.py]: Tried this, replaced `keep_sections` with `is_source`, but that did not work, unit-tests failed. Anyway, re-definition of the `compile()` method is necessary now, as it handles language code for ARB.

File: openformats/formats/json.py (hunk 3 above)
Comment (dsavinov-actionengine) [reply openformats/formats/json.py]: done

**Comment by dsavinov-actionengine** [comment]
> There is another issue I am concerned about. Looking at the examples of the [official Flutter documentation](https://docs.flutter.dev/ui/accessibility-and-internationalization/internationalization#setting-up)

...
>should we make sure that when is_source=False during compile, only the key-translation pairs should be present while when is_source=True we should make sure that they get back the same file they originally uploaded?
> Have you looked into this?

Did not implement this logic in the latest commit. It looks unusual if compared to other openformats handlers, which have their compiled files as similar as possible to the source files (only strings from stringset change). Maybe we could try to get back to this idea in a separate pull-request.

File: openformats/formats/json.py [hunk 4]
```diff
@@ -472,6 +479,156 @@ def unescape(string):
...
+            if key.startswith("@"):
+                continue
+            elif isinstance(value, (six.text_type)):
+                if not value.strip():
+                    continue
+                elif f"@{key}.type" in self.existing_keys:
+                    continue
+
+                context_key = f"@{key}.context"
+                context_value = self.metadata[context_key] \
```
Comment (dsavinov-actionengine) [inline openformats/formats/json.py:554]: Implemented issue 1 from QA spreadsheet

File: openformats/formats/json.py [hunk 5]
```diff
@@ -472,6 +479,156 @@ def unescape(string):
...
+                    order=openstring.order,
+                    pluralized=openstring.pluralized)
+            for openstring in stringset
+        ]
+        new_template = self._replace_translations(
+            template, fake_stringset, False
+        )
+        new_template = self._clean_empties(new_template)
+
+        if language_info is not None:
```
Comment (dsavinov-actionengine) [inline openformats/formats/json.py:609]: Implemented issue 2 from QA spreadsheet

File: openformats/formats/json.py [hunk 6]
```diff
@@ -472,6 +479,156 @@ def unescape(string):
...
+                if not value.strip():
+                    continue
+                elif f"@{key}.type" in self.existing_keys:
+                    continue
+
+                context_key = f"@{key}.context"
+                context_value = self.metadata[context_key] \
+                    if context_key in self.metadata.keys() else ""
+                description_key = f"@{key}.description"
+                description_value = self.metadata[description_key] \
```
Comment (dsavinov-actionengine) [inline openformats/formats/json.py:557]: Implemented issue 3 from QA spreadsheet

**Review by kbairak — APPROVED** [review]
